worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+	if !s.nextReadTriggered && len(s.remaining) < int(s.u.maxOverdrive) {
+		s.nextReadTriggered = true
+		s.u.triggerNextSlab(s.uploadID)
+	}
+
+	if len(s.remaining)%5 == 0 || len(s.remaining) < 5 {
+		fmt.Printf("DEBUG PJ: %v | %v | remaining sectors %d\n", s.uploadID, s.shardID, len(s.remaining))
+	}
+	return len(s.remaining) == 0
+}
+
+func (s *uploadState) inflight() uint64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.numInflight
+}
+
+func (s *uploadState) performance() float64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	downloaded := int64(s.numCompleted) * rhpv2.SectorSize
+	elapsed := time.Since(s.created)
+	return mbps(downloaded, elapsed.Seconds())
+}
+
+func (s *uploadState) overdriveCnt() int {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return int(s.numLaunched) - len(s.sectors)
+}
+
+func (s *uploadState) overdrivePct() float64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	numOverdrive := int(s.numLaunched) - len(s.sectors)
+	if numOverdrive <= 0 {
+		return 0
+	}
+
+	return float64(numOverdrive) / float64(len(s.sectors))
+}
+
+func (s *uploadState) overdrive(responseChan chan sectorResponse, shards [][]byte) *uploadJob {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// overdrive is not kicking in yet
+	if uint64(len(s.remaining)) >= s.u.maxOverdrive {
+		return nil
+	}
+
+	// overdrive is not due yet
+	if time.Since(s.lastOverdrive) < s.u.overdriveTimeout {
+		return nil
+	}
+
+	// overdrive is maxed out
+	if s.numInflight-uint64(len(s.remaining)) >= s.u.maxOverdrive {
+		return nil
+	}
+
+	// overdrive the remaining sector with the least number of overdrives
+	lowestSI := -1
+	s.overdriving[lowestSI] = math.MaxInt
+	for sI := range s.remaining {
+		if s.overdriving[sI] < s.overdriving[lowestSI] {
+			lowestSI = sI
+		}
+	}
+	if lowestSI == -1 {
+		return nil
+	}
+
+	fmt.Printf("DEBUG PJ: %v | %v | launching overdrive for sector %d\n", s.uploadID, s.shardID, lowestSI)
+	return &uploadJob{
+		requestCtx: s.remaining[lowestSI].ctx,
+
+		overdrive:    true,
+		responseChan: responseChan,
+
+		sectorIndex: lowestSI,
+		sector:      (*[rhpv2.SectorSize]byte)(shards[lowestSI]),
+		uploadID:    s.uploadID,
+		shardID:     s.shardID,
+	}
+}
+
+func (s *uploadState) cleanup() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for _, sCtx := range s.remaining {
+		sCtx.cancel()
+	}
+}
+
+func (s *uploadState) finish() ([]object.Sector, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	remaining := len(s.remaining)
+	if remaining > 0 {
+		return nil, fmt.Errorf("failed to upload slab: remaining=%d, inflight=%d, completed=%d launched=%d contracts=%d errors=%w", remaining, s.numInflight, s.numCompleted, s.numLaunched, s.u.numContracts(), s.errs)
+	}
+	return s.sectors, nil
+}
+
+func (id uploadID) String() string {
+	return fmt.Sprintf("%x", id[:])
+}
+
+func (a *dataPoints) average() float64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.avg
+}
+
+func (a *dataPoints) recompute() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// calculate sums
+	var data stats.Float64Data
```

Comment:
This was mostly a hack to get the percentiles. Maybe it can be cleaned up to make `dataPoints` a wrapper for `stats.Float64Data` instead of populating it in `recompute` like this.